documentation/debug.md

```diff
@@ -0,0 +1,150 @@
+		1. [ddd](#151-ddd)
+		2. [GNU Visual Debugger (gvd)](#152-gnu-visual-debugger-gvd)
+2. [Known issues](#2-known-issues)
+
+In this document we would like to describe how to debug OP-TEE. Depending on the
+platform you are using you will have a couple of different options.
+
+# 1. QEMU
+To debug OP-TEE when using QEMU you could use `gdb` as the main debugger. Using
+this setup will also make it possible to use some frontends for gdb if you don't
+feel comfortable using a command line debugging tool.
+
+## 1.1 Prerequisites
+Since there are inter-dependencies between the gits used when building OP-TEE,
+we recommend that you have been using the
+[setup_qemu_optee.sh](https://raw.githubusercontent.com/OP-TEE/optee_os/master/scripts/setup_qemu_optee.sh)
+script when setting up the environment for running OP-TEE using QEMU. In this
+guide we will use the default paths used in that particular script. I.e.
+
+```
+# Root folder for the project
+$HOME/devel/qemu_optee
+```
+
+
+## 1.2 Download gdb for ARM
+There are a variety of gdb binaries which claims to support ARM. We have
+experienced problems using some of them, therefore we recommend to use the
+`arm-none-eabi-gdb` that you will find built and hosted by Linaro. Therefore
+start by downloading this toolchain
+
+
+```
+wget https://releases.linaro.org/14.09/components/toolchain/binaries/gcc-linaro-arm-none-eabi-4.9-2014.09_linux.tar.xz
+```
+
+and put it in the toolchain folder
+
+
+```
+cd $HOME/devel/qemu_optee/toolchains
+tar xvf $DOWNLOAD_FOLDER/gcc-linaro-arm-none-eabi-4.9-2014.09_linux.tar.xz
+```
+
+## 1.3 Scripts
+A few helper scripts that makes life easier.
+
+Start by creating `$HOME/.gdbinit` and add:
+
+```
+set print array on
+set print pretty on
+
+define optee
+	handle SIGTRAP noprint nostop pass
+	symbol-file $HOME/devel/qemu_optee/optee_os/out/arm-plat-vexpress/core/tee.elf
+	target remote localhost:1234
+end
+document optee
+	Loads and setup the binary (tee.elf) for OP-TEE and also connects to the QEMU
+	remote.
+end
+```
+
+Now you are good to go for doing debugging using command line.
+
+## 1.4 Debug
+Start QEMU according to the instructions
+[here](https://github.com/OP-TEE/optee_os#444-boot-and-run-qemu-and-op-tee),
+*however*, do not start the emulation, i.e. do not type the `c` command in QEMU.
+The main reason for not doing so is because you cannot set the breakpoints on
+secure side when when kernel has booted up (if anyone knows why, please let us
+now about it, we haven't investigated it) and then in another shell start gdb
+like this:
+```
+$ $HOME/devel/qemu_optee/toolchains/gcc-linaro-arm-none-eabi-4.9-2014.09_linux/bin/arm-none-eabi-gdb -q
+```
+
+To connect to the remote and to load the `tee.elf`, simply type:
+```
+(gdb) optee
+SIGTRAP is used by the debugger.
+Are you sure you want to change it? (y or n) [answered Y; input not from
+terminal]
+0x00000000 in ?? ()
+```
+
+Now it is time to set the breakpoints. For example
+
+```
+(gdb) b plat_tee_entry 
```

Comment:
Travis / checkpatch complains about trailing space ... I'll fix that in next patch.